compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -730,26 +715,246 @@ fn encode_ty<'tcx>(
     typeid
 }
 
+/// Returns true if the FnAbi is an unpacked "rust-call" ABI closure call.
+fn is_closure_call<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> bool {
+    if fn_abi.conv == Conv::Rust
+        && !fn_abi.c_variadic
+        && !fn_abi.args.is_empty()
+        && let ty::Closure(_, args) = fn_abi.args[0].layout.ty.peel_refs().kind()
+    {
+        let fn_sig = tcx.signature_unclosure(args.as_closure().sig(), hir::Unsafety::Normal);
+        return fn_sig
+            .skip_binder()
+            .inputs()
+            .iter()
+            .eq_by(&fn_abi.args[1..], |ty, arg| *ty == arg.layout.ty)
+            && fn_sig.output().skip_binder() == fn_abi.ret.layout.ty;
+    }
+    false
+}
+
+/// Returns true if the FnAbi is a dynamic Fn trait call.
+fn is_dynamic_fn_call<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> bool {
+    if fn_abi.conv == Conv::Rust
+        && !fn_abi.c_variadic
+        && !fn_abi.args.is_empty()
+        && let ty::Dynamic(predicates, ..) = fn_abi.args[0].layout.ty.peel_refs().kind()
+    {
+        if predicates.is_fn_trait(tcx) {
+            return predicates
+                .fn_trait_args(tcx)
+                .unwrap()
+                .iter()
+                .eq_by(&fn_abi.args[1..], |ty, arg| ty == arg.layout.ty)
+                && predicates.fn_trait_output(tcx).unwrap() == fn_abi.ret.layout.ty;
+        }
+    }
+    false
+}
+
+/// Returns true if the trait object is an Fn trait object.
+fn is_dynamic_fn_trait<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> bool {
+    match ty.kind() {
+        ty::Dynamic(predicates, ..) => predicates.is_fn_trait(tcx),
+        _ => false,
+    }
+}
+
+/// Creates a trait object.
+fn new_dynamic_trait<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    args: GenericArgsRef<'tcx>,
+) -> Ty<'tcx> {
+    let predicate =
+        ty::ExistentialPredicate::Trait(ty::ExistentialTraitRef { def_id: def_id, args: args });
+    let predicates = tcx.mk_poly_existential_predicates(&[ty::Binder::dummy(predicate)]);
+    Ty::new_dynamic(tcx, predicates, tcx.lifetimes.re_erased, ty::Dyn)
+}
+
+/// Creates an Fn trait object.
+fn new_dynamic_fn_trait<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    args: &[Ty<'tcx>],
+    output: Ty<'tcx>,
+) -> Ty<'tcx> {
+    let trait_ref = ty::ExistentialPredicate::Trait(ty::ExistentialTraitRef {
+        def_id: def_id,
+        args: tcx.mk_args(&[Ty::new_tup(tcx, args).into()]),
+    });
+    // HACK(rcvalle): It is okay to to use the DefId of the TraitRef here because it is never used
```

Comment:
I don't think this should be done. It should be possible to acquire the `FnOnce::Output` def id to construct a proper existential projection predicate here.

Reply:
Done.